src/btree/bt_cursor.c

```diff
@@ -232,107 +232,195 @@ __cursor_fix_implicit(WT_BTREE *btree, WT_CURSOR_BTREE *cbt)
+
+        /* The key value pair we were trying to return weren't within the given bounds. */
+        if (key_out_of_bounds)
+            return (0);
+    }
+
+    WT_RET(__wt_txn_read_upd_list(session, cbt, cbt->ins->upd));
+    if (cbt->upd_value->type != WT_UPDATE_INVALID) {
+        if (cbt->upd_value->type == WT_UPDATE_TOMBSTONE)
+            return (0);
+        *valid = true;
+    }
+
+    return (0);
+}
+
+/*
+ *  __cursor_valid_int_ondisk --
+ *     Check the ondisk value or history store for a valid update.
+ */
+static int
+__cursor_valid_int_ondisk(WT_CURSOR_BTREE *cbt, WT_ITEM *key, WT_UPDATE *upd, bool *valid)
+{
+    WT_SESSION_IMPL *session;
 
-    btree = CUR2BT(cbt);
-    page = cbt->ref->page;
     session = CUR2S(cbt);
-    upd = NULL;
 
     /*
-     * We may be pointing to an insert object, and we may have a page with
-     * existing entries.  Insert objects always have associated update
-     * objects (the value).  Any update object may be deleted, or invisible
-     * to us.  In the case of an on-page entry, there is by definition a
-     * value that is visible to us, the original page cell.
-     *
-     * If we find a visible update structure, return our caller a reference
-     * to it because we don't want to repeatedly search for the update, it
-     * might suddenly become invisible (imagine a read-uncommitted session
-     * with another session's aborted insert), and we don't want to handle
-     * that potential error every time we look at the value.
+     * Check for a value on disk or in the history store, passing in any update.
      *
-     * Unfortunately, the objects we might have and their relationships are
-     * different for the underlying page types.
+     * Potentially checking an update chain twice (both here, and above if the insert list is set in
+     * the case of column-store), isn't a mistake. In a modify chain, if rollback-to-stable recovers
+     * a history store record into the update list, the base update may be in the update list and we
+     * must use it rather than falling back to the on-disk value as the base update.
+     */
+    WT_RET(__wt_txn_read(session, cbt, key, cbt->recno, upd));
+    if (cbt->upd_value->type != WT_UPDATE_INVALID) {
+        if (cbt->upd_value->type == WT_UPDATE_TOMBSTONE)
+            return (0);
+        *valid = true;
+    }
+    return (0);
+}
+
+/*
+ *  __cursor_valid_int_row --
+ *     Determine if a valid update is being referenced by the cursor for row-store.
+ */
+static int
+__cursor_valid_int_row(WT_CURSOR_BTREE *cbt, bool *valid, bool check_bounds)
+{
+    WT_DECL_ITEM(key);
+    WT_PAGE *page;
+    WT_SESSION_IMPL *session;
+    WT_UPDATE *upd;
+    bool key_out_of_bounds;
+
+    session = CUR2S(cbt);
+    upd = NULL;
+    page = cbt->ref->page;
+    key_out_of_bounds = false;
+
+    /* For all row-store cases we can extract the key from cbt->tmp if the compare value is zero. */
+    if (cbt->compare == 0 && cbt->tmp != NULL)
```

Comment:
This as far as I am aware is the only "risky" part of the refactor, when I pulled this logic into the function it was correct for all the callers of `__wt_cursor_valid` but in theory newer calls may not honour this. I'm not sure if I can handle them calling it in a "bad" way. If this is confusing let me know.

Reply:
@agorrod this also feels a bit like a layering issue? I couldn't think of a better way to avoid passing keys but realistically the caller would know better what key to pass rather than assuming it's valid based off the compare value. Its cleaner but does feel a bit odd.

Reply:
I wonder if the cbt->tmp buffer is what causes this confusion, perhaps we should just have cursor->key. But I imagine it was added for a reason that I will encounter when I try to remove it.

Reply:
I think layering wise this is OK here, and it's better to have the assumption noted here in one place than distributed around.
I have to confess to not fully understanding cbt->tmp (a different name would help), but I don't think we should open that potential can of worms here.